src/System.Diagnostics.EventLog/tests/EventLogTests/EventLogWriteEntryTests.cs

```diff
@@ -0,0 +1,375 @@
+        {
+            string source = "Source" + nameof(WriteEntryWithTypeIdAndCategory);
+            string log = "CategoryEntry";
+            try
+            {
+                EventLog.CreateEventSource(source, log);
+                EventLogEntry eventLogEntry;
+                if (sourceFlag)
+                    eventLogEntry = WriteLogEntry(source, type: true, instance: true, category: true);
+                else
+                    eventLogEntry = WriteLogEntryWithSource(source, type: true, instance: true, category: true);
+
+                //There is some prefix string already attached to the message passed
+                Assert.Contains(message, eventLogEntry.Message);
+                Assert.Equal((short)eventInstance.CategoryId, eventLogEntry.CategoryNumber);
+                Assert.Equal("(" + eventLogEntry.CategoryNumber + ")", eventLogEntry.Category);
+            }
+            finally
+            {
+                EventLog.DeleteEventSource(source);
+                EventLog.Delete(log);
+            }
+        }
+
+        [ConditionalTheory(typeof(Helpers), nameof(Helpers.IsElevatedAndSupportsEventLogs))]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void WriteEntryWithTypeIdCategoryAndData(bool sourceFlag)
+        {
+            string source = "Source" + nameof(WriteEntryWithTypeIdCategoryAndData);
+            string log = "EntryData";
+            try
+            {
+                EventLog.CreateEventSource(source, log);
+                EventLogEntry eventLogEntry;
+                if (sourceFlag)
+                    eventLogEntry = WriteLogEntry(source, type: true, instance: true, category: true, data: true);
+                else
+                    eventLogEntry = WriteLogEntryWithSource(source, type: true, instance: true, category: true, data: true);
+
+                Assert.Contains(message, eventLogEntry.Message);
+                Assert.Equal(rawData, eventLogEntry.Data);
+            }
+            finally
+            {
+                EventLog.DeleteEventSource(source);
+                EventLog.Delete(log);
+            }
+        }
+
+        [ConditionalTheory(typeof(Helpers), nameof(Helpers.IsElevatedAndSupportsEventLogs))]
+        public void WriteEntryWithoutSource()
+        {
+            using (EventLog eventLog = new EventLog())
+            {
+                Assert.Throws<ArgumentException>(() => eventLog.WriteEntry(message));
+            }
+
+        }
+
+        [ConditionalTheory(typeof(Helpers), nameof(Helpers.IsElevatedAndSupportsEventLogs))]
+        public void WriteEntryWithInvalidType()
+        {
+            using (EventLog eventLog = new EventLog())
+            {
+                string source = "Source_" + nameof(WriteEntryWithInvalidType);
+                eventLog.Source = source;
+                Assert.Throws<InvalidEnumArgumentException>(() => eventLog.WriteEntry(message, (EventLogEntryType)7));
```

Comment:
don't cast a number to EventLogEntryType at runtime. Use the language -- EventLogEntryType.XXX | EventLogentryType.YYY. Then it's checked compiletime

Reply:
7 is a random number which is not associated with any type in EventLogEntryType.
If we change it to EventLogEntryType.XXX , it will give a compile error and we wont be able to test if the system throws this particular exception or not